### DISCUSSION TITLE
Docker compose constraints

### DISCUSSION THREAD

**Discussion by aakarim** [cue-lang/cue#1280] [Q&A] [answered]
Are there any existing constraints for docker compose files?

**Comment by verdverm** [comment]
Can you provide more context or what you are trying to achieve? Are you referring to resource (cpu,mem) settings?

**Reply by aakarim** [reply]
We have pre-built constraints/imports for Kubernetes manifest files, to ensure the resulting exports are correct. I was just wondering whether there is something similar for Docker Compose files. 

```
version: "3.8"
services: [string]: compose.#Service
```
or along those lines.

**Comment by slewiskelly** [comment]
Do you want a CUE representation of the [defined JSON schema](https://github.com/compose-spec/compose-spec/blob/master/schema/compose-spec.json)?

There seems to be a [Go representation](https://pkg.go.dev/github.com/compose-spec/compose-go/types) which I was able to `cue get`, but I'm not sure if this is what was desired:

```console
$ go get github.com/compose-spec/compose-go/types
$ cue get go github.com/compose-spec/compose-go/types
$ tree cue.mod/gen/
cue.mod/gen/
└── github.com
    └── compose-spec
        └── compose-go
            └── types
                ├── config_go_gen.cue
                ├── project_go_gen.cue
                └── types_go_gen.cue

4 directories, 3 files
```

**Comment by verdverm** [comment, marked as answer]
@aakarim You can import any JSON Schema into CUE with the `cue import` command. As @slewiskelly pointed out, one already exists for docker-compose. You can find many more on the [Schema Store](https://www.schemastore.org/json/)

I wrote the following tutorial: https://cuetorials.com/first-steps/convert-jsonschema/

**Reply by aakarim** [reply]
Thank you!
